hca-to-scea/helpers/utils.py

```diff
@@ -0,0 +1,260 @@
+            pooled_samples_cell_line = [cell_line_id for cell_line_id in cell_line_ids if "||" in cell_line_id]
+
+    pooled_samples_organoid = []
+    if "organoid" in xlsx_dict.keys() and not xlsx_dict["organoid"].empty:
+        df = xlsx_dict["organoid"]
+        if 'organoid.biomaterial_core.biomaterial_id' in df.columns:
+            organoid_ids = list(df['organoid.biomaterial_core.biomaterial_id'])
+            pooled_samples_organoid = [organoid_id for organoid_id in organoid_ids if "||" in organoid_id]
+
+    df = xlsx_dict["donor_organism"]
+    donor_ids = list(df['donor_organism.biomaterial_core.biomaterial_id'])
+    pooled_samples_donor = [str(donor_id) for donor_id in donor_ids if "||" in str(donor_id)]
+
+    if pooled_samples_specimen or pooled_samples_donor or pooled_samples_cell_line or pooled_samples_organoid:
+        pooled_samples = True
+    else:
+        pooled_samples = False
+
+    return pooled_samples
+
+def filter_biomaterials(xlsx_dict, xlsx_dict_tmp, library_protocol):
+
+    xlsx_dict_tmp["sequence_file"] = xlsx_dict["sequence_file"].loc[
+        xlsx_dict["sequence_file"]["library_preparation_protocol.protocol_core.protocol_id"] == library_protocol]
+    cell_suspension_ids = xlsx_dict_tmp["sequence_file"]["cell_suspension.biomaterial_core.biomaterial_id"].values
+    xlsx_dict_tmp["cell_suspension"] = xlsx_dict["cell_suspension"][
+        xlsx_dict["cell_suspension"]["cell_suspension.biomaterial_core.biomaterial_id"].isin((cell_suspension_ids))]
+
+    biomaterial_tabs = ["specimen_from_organism", "organoid", "cell_line"]
+    for biomaterial_tab in biomaterial_tabs:
+        if biomaterial_tab in xlsx_dict.keys():
+            if not xlsx_dict_tmp[biomaterial_tab].empty:
+                id_column = "{}.biomaterial_core.biomaterial_id".format(biomaterial_tab)
+                if id_column in xlsx_dict_tmp["cell_suspension"].columns:
+                    if not xlsx_dict_tmp["cell_suspension"][id_column].empty and not xlsx_dict_tmp["cell_suspension"][id_column].isna().all():
+                        specimen_ids = xlsx_dict_tmp["cell_suspension"][
+                            "{}.biomaterial_core.biomaterial_id".format(biomaterial_tab)].values
+                        xlsx_dict_tmp[biomaterial_tab] = xlsx_dict[biomaterial_tab][
+                            xlsx_dict[biomaterial_tab]["{}.biomaterial_core.biomaterial_id".format(biomaterial_tab)].isin(
+                                (specimen_ids))]
+                if "donor_organism.biomaterial_core.biomaterial_id" in xlsx_dict_tmp[biomaterial_tab].columns:
+                    donor_ids_specimen = xlsx_dict_tmp[biomaterial_tab][
+                        "donor_organism.biomaterial_core.biomaterial_id"].values
+                    xlsx_dict_tmp["donor_organism"] = xlsx_dict["donor_organism"][
+                        xlsx_dict["donor_organism"]["donor_organism.biomaterial_core.biomaterial_id"].isin(
+                            (donor_ids_specimen))]
+
```

Comment:
Too many nesting levels and branching.
This is untestable and will become unreadable, even by the author, in no time.

Reply:
I thought this too :-/ but I am not currently sure of a better way to do what I need the function to do. For now, I would like to leave it so it can be merged into master. And come back to it later. This is because I want us to be able to use the latest version of the tool, with the new functionality.